libtenzir/include/tenzir/pipeline.hpp

```diff
@@ -154,6 +154,42 @@ struct inspector
   }
 };
 
+// Metrics that track the information about inbound and outbound elements that
+// pass through this operator.
+struct [[nodiscard]] pipeline_op_metrics {
+  size_t index = 0;
+  caf::timespan time_starting = {};
+  caf::timespan time_running = {};
+  caf::timespan time_scheduled = {};
+  caf::timespan time_elapsed = {};
+  std::string in_unit = {};
+  std::string out_unit = {};
+  uint64_t inbound_total = {};
+  uint64_t inbound_num_batches = {};
+  double inbound_rate_per_second = {};
+  uint64_t outbound_total = {};
+  uint64_t outbound_num_batches = {};
+  double outbound_rate_per_second = {};
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, pipeline_op_metrics& x) -> bool {
+    return f.object(x)
+      .pretty_name("pipeline_op_metrics")
+      .fields(f.field("index", x.index),
+              f.field("time_starting", x.time_starting),
+              f.field("time_running", x.time_running),
+              f.field("time_scheduled", x.time_scheduled),
+              f.field("time_elapsed", x.time_elapsed),
+              f.field("in_unit", x.in_unit), f.field("out_unit", x.out_unit),
+              f.field("inbound_total", x.inbound_total),
+              f.field("inbound_num_batches", x.inbound_num_batches),
+              f.field("inbound_rate_per_second", x.inbound_rate_per_second),
+              f.field("outbound_total", x.outbound_total),
+              f.field("outbound_num_batches", x.outbound_num_batches),
+              f.field("outbound_rate_per_second", x.outbound_rate_per_second));
+  }
+};
```

Comment:
I have one small change request: Given that we may want to add a toggle between events and estimated bytes, we may have values for both bytes and events on either side in the future. So it may make sense to model the data more like this:

```c++
struct measurement {
  std::string unit; // "events" or "bytes"
  uint64_t num_total;
  uint64_t num_batches;
};

struct operator_metric {
  std::vector<measurement> inbound_measurements;
  std::vector<measurement> outbound_measurements;
  uint64_t index;
  duration time_starting;
  duration time_running;
  duration time_scheduled;
  duration time_elapsed;
};
```

There's also a small bug fix: we should not use `size_t` here, because it is platform-specific. This can lead to subtle incompatibilities between platforms.

The rate can also be easily calculated at the receiving end, so I don't think we should include it in the data itself.

Reply:
Instead of a `unit` filed I'd rather have
```c++
struct measurement {
  uint64_t num_events;
  uint64_t num_bytes;
  uint64_t num_batches;
};
```
I think an approximate value for `num_bytes` is fine and shouldn't be hard to calculate. For operators that deal in chunks we set `num_events` to 0.

Reply:
Looking forward we also want to have series as an element type, so then you'd have every single type we have as a possible unit. That's why I think we want to have the more generic solution.

Reply:
I agree with all points, but:

>The rate can also be easily calculated at the receiving end, so I don't think we should include it in the data itself.

I experimented with that and realized that the `rate` values can slightly differ from where the time endpoint is generated. The time endpoint when an execution node is finished/hits the metrics emission loop would be slightly different from the time endpoint when a pipeline has stopped in the pipeline manager.

When calculating metrics in the execution engine, I would prefer to calculate metrics as closest inside the execution engine as possible, then pass it to other recipients, hence the `rate` fields. WDYT?

Reply:
Tobias and I just agreed on the following in a quick call:

```cpp
struct measurement {
  std::string unit;
  uint64_t num_elements;
  uint64_t num_batches;

  // Always 0 for now for events until we find a way
  // to reliably measure this and always the same as
  // num_elements for bytes.
  uint64_t num_approx_bytes;
};

struct metric {
  uint64_t operator_index;

  measurement inbound_measurement;
  measurement outbound_measurement;

  duration time_starting;
  duration time_running;
  duration time_scheduled;
  duration time_elapsed; 
};
```

Reply:
> I experimented with that and realized that the rate values can slightly differ depending on where the time endpoint is generated. The time endpoint when an execution node is finished/hits the metrics emission loop would be slightly different from the time endpoint when a pipeline has stopped in the pipeline manager.

I think the rate should be the same as `{in,out}bound_measurement.num_elements / std::chrono::duration_cast<std::chrono::duration<double, std::chrono::seconds::period>>(time_elapsed)` given the data from the struct I posted. That's exactly what the execution node uses to calculate it.

Reply:
OK, then the `rate` calculation would be easier than I suspected after all. I can remove the `rate`s then.

Reply:
I split up `pipeline_op_metrics` in https://github.com/tenzir/tenzir/pull/3376/commits/4d9a7e8c27e1a57b626ff75ff915fc38c7cb9890